### ISSUE TITLE
Question about SoftSilhouetteShader output vs binary GT masks for loss

### ISSUE THREAD

**Issue by Frklin** [facebookresearch/pytorch3d#2015]
Hi, and thanks for PyTorch3D!

I have a question about how to properly use the silhouette renderer for optimization.

**Setup**
I’m using SoftSilhouetteShader with very small sigma/gamma because I was hoping to get (almost) binary silhouettes:
```
sigma = 1e-4
blend_params = BlendParams(
    sigma=sigma,
    gamma=1e-4,
    background_color=color_bg,
)

raster_settings_silhouette = RasterizationSettings(
    image_size=image_size,
    blur_radius=np.log(1.0 / 1e-4 - 1.0) * sigma,
    faces_per_pixel=50,
)

# silhouette renderer
renderer_silh = MeshRenderer(
    rasterizer=MeshRasterizer(
        cameras=cameras,
        raster_settings=raster_settings_silhouette,
    ),
    shader=SoftSilhouetteShader(blend_params=blend_params),
)

silh_images = renderer_silh(mesh)[..., 3]  # (B, H, W), values in [0, 1]
```
When I visualize silh_images, the result is a smooth silhouette:
values can be 0 and 1, but there are also many intermediate values like 0.23, 0.88, etc., especially around object boundaries. I also noticed that the rendered silhouette contains small holes, especially in regions like the eyes or thin/concave areas. In this image is possible to see the render of the mesh (on the left) and its silhouette (on the right)

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/f67ccb7d-a040-4b97-a54e-e23cdd45613a" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/01a483b6-d348-47e6-a2e9-2b1b9d0f23ba" />

#### What I expected
I was expecting something closer to a binary mask with only 0s and 1s (or at least very close to that with these tiny sigma/gamma values) since I thought the silhouette would simply indicate whether each pixel was hit by any part of the mesh.

**How I’m using it in optimization**
In my optimization loop I have a ground-truth silhouette mask with values strictly in {0, 1} and I compute a loss like this:
```
loss_sil = torch.nn.SmoothL1Loss()(silh_images, gt_mask)
```
My doubt is:
- The rendered silhouette is a soft mask (continuous in [0, 1]),
- The ground truth is a binary mask (0/1 only),

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/01a483b6-d348-47e6-a2e9-2b1b9d0f23ba" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/7c6816a7-e1a2-4c81-98a2-c6ae08e6278f" />

so I’m not sure if directly comparing these with SmoothL1Loss is the intended / recommended way to use SoftSilhouetteShader for optimization.

I care about **gradients**, so I’d like to avoid hard thresholding (silh_images > 0.5).float() inside the computation graph, since that would make the operation non-differentiable almost everywhere.

### Questions:
1) Is it expected behavior that SoftSilhouetteShader produces soft values in (0,1) even when sigma and gamma are as small as 1e-4?

2) For optimization, is it correct / recommended to compare this soft silhouette to a binary ground-truth mask using a regression-style loss (e.g. SmoothL1Loss, L1, BCE), or should I be doing some additional transformation (e.g. sharpening, different loss, etc.)?

3) If the goal is to get something visually close to a hard silhouette but still differentiable and suitable for gradient-based optimization, is there an alternative approach (e.g. using pix_to_face + a differentiable surrogate)?

## Environment

PyTorch3D version: 0.7.8
PyTorch version: 2.5.1
CUDA version: 12.1

**Comment by Frklin** [comment]
Thanks for the reply, it seems that with 1e-5 the holes disappeared (I tried with bin_size=0 and sigma=1e-4, however the holes did remain)
